### PR TITLE
Calculate upload rate

### PR DESCRIPTION
**Pull request by gilest**
Continuation of work by @RobbieTheWagner  in #1015

### Methodology

- Record rate data against each file as upload `progress` events occur
- Calculate weighted moving average of rate per-file as `File.rate` property
- Sum all uploading files as `Queue.rate` property

All rates are in bytes/ms.

### Demo

Recent improvements to the demo page make this much easier to validate and test.

https://github.com/adopted-ember-addons/ember-file-upload/assets/36919/6c35b35b-93a5-4368-91c6-0cc301730ffb

### CODE REVIEW

File: ember-file-upload/src/system/upload.ts [hunk 1]
```diff
@@ -52,6 +52,28 @@ function normalizeOptions(
...
+    const timeElapsedSinceLastUpdate =
+      updatedTime - file.timestampWhenProgressLastUpdated;
+
+    const bytesTransferredSinceLastUpdate =
+      file.loaded - file.bytesWhenProgressLastUpdated;
+
+    // Divide by elapsed time to get bytes per millisecond
+    const rate = bytesTransferredSinceLastUpdate / timeElapsedSinceLastUpdate;
+
+    file.rates = [...file.rates, rate];
```
Comment (gilest) [inline ember-file-upload/src/system/upload.ts:69]: Would it be more performant to use a `TrackedArray`?

File: ember-file-upload/src/upload-file.ts [hunk 2]
```diff
@@ -51,6 +52,11 @@ export class UploadFile {
     this.#name = value;
   }
 
+  /** The current speed in ms that it takes to upload one byte */
+  get rate() {
+    return estimatedRate(this.rates);
```
Comment (gilest) [inline ember-file-upload/src/upload-file.ts:57]: Work done in plain functions for unit-testing purposes.

**Comment by gilest** [comment]
Profiling with 6x CPU slowdown the `estimatedRate` function takes ~1ms at most

image

Comparison to date-fns `intervalToDuration` in the same period
image

File: ember-file-upload/src/system/upload.ts (hunk 1 above)
Comment (RobbieTheWagner) [reply ember-file-upload/src/system/upload.ts:69]: I'm not sure. I say leave it unless we have problems.

**Pull request merged**